src/inputs/TreeSelectField/TreeSelectField.tsx

```diff
@@ -157,29 +164,53 @@ function TreeSelectFieldBase<O, V extends Value>(props: TreeSelectFieldProps<O,
 
   const { collapsedKeys } = useTreeSelectFieldProvider();
 
-  function levelOptions(o: NestedOption<O>, level: number, filtering?: boolean): LeveledOption<O>[] {
-    // If a user is filtering, then do not provide level to the options as the various paddings may look quite odd.
-    const actualLevel = filtering ? 0 : level;
-    return [
-      [o, actualLevel],
-      ...(o.children?.length && !collapsedKeys.includes(valueToKey(getOptionValue(o)))
-        ? o.children.flatMap((oc: NestedOption<O>) => levelOptions(oc, actualLevel + 1, filtering))
-        : []),
-    ];
-  }
+  const levelOptions = useCallback(
+    (o: NestedOption<O>, level: number, filtering?: boolean): LeveledOption<O>[] => {
+      // If a user is filtering, then do not provide level to the options as the various paddings may look quite odd.
+      const actualLevel = filtering ? 0 : level;
+      return [
+        [o, actualLevel],
+        ...(o.children?.length && !collapsedKeys.includes(valueToKey(getOptionValue(o)))
+          ? o.children.flatMap((oc: NestedOption<O>) => levelOptions(oc, actualLevel + 1, filtering))
+          : []),
+      ];
+    },
+    [collapsedKeys, getOptionValue],
+  );
 
-  // Initialize the TreeFieldState
-  const [fieldState, setFieldState] = useState<TreeFieldState<O>>(() => {
+  // Given an option and the selected keys, it will return the top most selected option in the option's tree.
+  // This could be a parent of the option, or the option itself.
+  const getTopLevelSelections = useCallback(
+    (o: NestedOption<O>, selectedKeys: Set<Key>): NestedOption<O>[] => {
+      // If this element is already selected, return early. Do not bother looking through children.
+      if (selectedKeys.has(valueToKey(getOptionValue(o)))) return [o];
+      // If this element has no children, then look through the children for top level selected options.
+      if (o.children) return [...o.children.flatMap((c) => getTopLevelSelections(c, selectedKeys))];
+      return [];
+    },
+    [getOptionValue],
```

Comment:
I.e. for this function, if we added `getOptionValue` as a param, it could be a top-level function at the bottom of the file. Granted, callers would have to a) pass `getOptionValue` and b) add `getOptionValue` to their own deps array (if applicable) but that doesn't seem too bad.

Like if there were ~4-5+ deps here, then maybe it's more worth it, to let the caller just depend on the `useCallback`-d function and not know/care what it's non-parameter deps / deps-that-come-from-props are.